### PR TITLE
Closes #1470: Add support for Authentication dialogs.

### LINKED ISSUE #1470: Add support for  authentication alerts for GeckoEngine
User Story: https://github.com/mozilla-mobile/reference-browser/issues/223

### PR DESCRIPTION
**Pull request by Amejia481**
[Demo](https://drive.google.com/file/d/1S_nRv8Pu-GKnzLCCVRiFiqYTSU-40feC/view)

You can test with this [website](http://the-internet.herokuapp.com/basic_auth) admin/admin

### CODE REVIEW

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt [hunk 1]
```diff
@@ -82,4 +82,44 @@ sealed class PromptRequest {
...
+     * For more related info take a look at
+     * <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication>MDN docs</a>
+     * @property title of the dialog.
+     * @property message the body of the dialog.
+     * @property userName default value provide for this session.
+     * @property password default value provide for this session.
+     * @property method type of authentication,  valid values [Method.HOST] and [Method.PROXY].
+     * @property level indicates the level of security of the authentication like [Level.NONE],
+     * [Level.SECURED] and [Level.PASSWORD_ENCRYPTED].
+     * @property onlyShowPassword indicates if the dialog only include a password field.
```
Comment (csadilek) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt]: nit: should only include a password field

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt [hunk 2]
```diff
@@ -82,4 +82,44 @@ sealed class PromptRequest {
...
+     * @property previousFailed indicates if this request is the result of a previous failed attend to login.
```
Comment (csadilek) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt]: nit: s/attent/attempt

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/AuthenticationDialogFragment.kt [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) = Unit
+
+            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) = Unit
+        })
+    }
+
+    companion object {
+        /**
+         * A builder method for creating a [AuthenticationDialogFragment]
+         * @param sessionId to create the dialog.
```
Comment (csadilek) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/AuthenticationDialogFragment.kt]: nit: ... the id of the session for which this dialog will be created...

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt [hunk 4]
```diff
@@ -297,6 +301,23 @@ class PromptFeature(
         }
     }
 
+    /**
+     * Invoked when the user requested to start the authentication flow.
+     * This consumes the [PromptFeature] value from the [Session] indicated by [sessionId].
+     * @param sessionId the id for which session the user doesn't want to show more dialogs.
```
Comment (csadilek) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt]: this line needs an update.

File: components/feature/prompts/src/test/java/mozilla/components/feature/prompts/AuthenticationDialogFragmentTest.kt [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+        dialog.show()
+
+        val usernameEditText = dialog.findViewById<EditText>(id.username)
+
+        assertEquals(usernameEditText.visibility, GONE)
+    }
+
+    @Test
+    fun `dialog with not title must has default value`() {
```
Comment (csadilek) [inline components/feature/prompts/src/test/java/mozilla/components/feature/prompts/AuthenticationDialogFragmentTest.kt]: nit: dialog without a title?

File: components/feature/prompts/src/test/java/mozilla/components/feature/prompts/AuthenticationDialogFragmentTest.kt [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+        usernameEditText.setText("new_username")
+        passwordEditText.setText("new_password")
+
+        assertEquals(usernameEditText.text.toString(), "new_username")
+        assertEquals(passwordEditText.text.toString(), "new_password")
+    }
+
+    @Test
+    fun `dialog with onlyShowPassword equals true must not have a username field`() {
```
Comment (csadilek) [inline components/feature/prompts/src/test/java/mozilla/components/feature/prompts/AuthenticationDialogFragmentTest.kt]: ~~equals true~~

File: components/feature/prompts/src/test/java/mozilla/components/feature/prompts/AuthenticationDialogFragmentTest.kt (hunk 5 above)
Comment (Amejia481) [reply components/feature/prompts/src/test/java/mozilla/components/feature/prompts/AuthenticationDialogFragmentTest.kt]: The title of the dialog can be optional and we have to provide a default value for that.
What about improving the test name to:

`when the title is not provided the dialog must has a default value`?

**Pull request merged**